content/blog/2025-05-08-go-client-GA.md

```diff
@@ -0,0 +1,501 @@
+	txs := []string{"Ping - 'Hello Valkey Cluster'"}
+
+	opts := pipeline.NewClusterBatchOptions().
+		WithRoute(config.RandomRoute)
+
+	// Execute transaction
+	results, err := clusterClient.ExecWithOptions(context.Background(), *clusterTx, false, *opts)
+	if err != nil {
+		log.Printf("Cluster transaction failed: %v", err)
+		return
+	}
+
+	// Process results
+	for i, result := range results {
+		fmt.Printf("Command %s results: %v\n", txs[i], result)
+	}
+```
+
+## Batch design
+
+The batch implementation in Valkey GLIDE was designed with several key goals, each achieved through specific implementation approaches:
+
+### Idiomatic Go API
+
+The design provides a natural, Go-like interface for batches through:
+
+- Clean command calling syntax (`tx.Set(...)`, `tx.Get(...)`)
+- Use of Go's embedding to inherit behavior from `BaseBatch`
+- Explicit error handling that follows Go conventions (`result, err := pipeline.Exec()`)
+- Use of method chaining to efficiently build command sequences (e.g. `tx.Set(...).Set(...).SAdd(...)`)
+- Use of Go context for timeout and cancellation support 
+
+### Command Transparency
+
+A important principle in the design of the client was transparently accessing the command, in both regular and batch contexts. While the server-side command execution is the same, the client-side result and error handling will differ since the batch results must account for multiple commands and potential errors. Go context is passed when the command is executed: with regular clients this occurs when the command is called, and in batches this occurs when the batch is executed. In batches, Cluster mode and standalone mode commands are the same with only a few exceptions for Select, Move, Scan, and PubSub commands. This is slightly different from regular contexts where some commands may have specific routing options that are available when using the ClusterClient.
+
+```go
+// In regular client context
+client := glide.NewClient(...)
+// Executes immediately with context
+client.Set(context.Background(), "key1", "value1")
+
+// In transaction context 
+tx := pipeline.NewStandaloneBatch(true)
+// Queued for later execution
+tx.Set("key1", "value1")
+// Execute transaction with context
+results, err := client.Exec(context.Background(), *tx, false)
+```
+
+### Optimistic Concurrency Support
+
+Valkey's optimistic locking mechanism is supported through:
+
+- Implementation of `Watch()` that monitors keys for changes,
+- Implementation of `Unwatch()` that releases monitoring,
+- Proper handling of transaction failure when watched keys change,
+- Passing watch state through to the Valkey server during execution.
+
+This enables patterns like:
+
+```go
+client.Watch(context.Background(), []string{"key1"})
+tx := pipeline.NewStandaloneBatch(true).
+    Get("key1").
+    Set("key1", "new-value")
+
+// Will fail if key1 changed between Watch and Exec
+raiseOnError := true
+results, err := client.Exec(context.Background(), *tx, raiseOnError)
+```
+
```

Comment:
Proposal: show example with context, pubsub, open telemetry

Reply:
I'm trying not to expand the blog post even further than it already is. As blog posts go, this is already quite large.  We should probably create separate blog posts for each of thos concepts, so that they can be adequately addressed. Quite honestly, if I were writing the posts from scratch, I would have created a general high level blog for the 2.0 release, and then separate blogs for each of the major features we want to highlight.  I'll let Madelyn and Kyle guide us on how they want to proceed here.

Reply:
Yeah, my opinion is that this post is already like 2x-3x longer than I would like. The average person reads a Valkey.io blog for about 3.5 minutes, so ideally we want it to be about that long and convey the N most important things. There is also a lot of overlap with https://valkey.io/blog/2025-03-4-go-client-in-public-preview/.

I think ideally I would just focus on batch support and open telemetry integration. Everything else seems duplicative of work discussed elsewhere.